### ISSUE TITLE
React-Native fetch, Network request failed. On android

### ISSUE THREAD

**Issue by Clcll** [facebook/react-native#10404]
Hello, everything working fine on IOS but it's faild on Android  

fetch('http://somesite.com/app/connect', { ........

Thank you before hand

**Comment by christopherdro** [comment]
@facebook-github-bot label Android

**Comment by dunhuang** [comment]
@Clcll please show the options of your fetch

**Comment by woniesong92** [comment]
My requests using `axios` are also failing:

```

axios.get(`http://localhost:5000/posts`)
  .then(response => {
    console.log(response)
  })
  .catch(error => {
    console.log(error)
  })
```

And I get the following exception on browser:
```
Error: Network Error
    at createError (http://localhost:8081/index.android.bundle?platform=android&dev=true&hot=true&minify=false:84821:11)
    at XMLHttpRequest.handleError (http://localhost:8081/index.android.bundle?platform=android&dev=true&hot=true&minify=false:84693:8)
    at XMLHttpRequest.dispatchEvent (http://localhost:8081/index.android.bundle?platform=android&dev=true&hot=true&minify=false:15132:15)
    at XMLHttpRequest.setReadyState (http://localhost:8081/index.android.bundle?platform=android&dev=true&hot=true&minify=false:30776:6)
    at XMLHttpRequest.__didCompleteResponse (http://localhost:8081/index.android.bundle?platform=android&dev=true&hot=true&minify=false:30622:6)
    at http://localhost:8081/index.android.bundle?platform=android&dev=true&hot=true&minify=false:30716:52
    at RCTDeviceEventEmitter.emit (http://localhost:8081/index.android.bundle?platform=android&dev=true&hot=true&minify=false:14371:23)
    at MessageQueue.__callFunction (http://localhost:8081/index.android.bundle?platform=android&dev=true&hot=true&minify=false:12115:34)
    at http://localhost:8081/index.android.bundle?platform=android&dev=true&hot=true&minify=false:11986:8
    at guard (http://localhost:8081/index.android.bundle?platform=android&dev=true&hot=true&minify=false:11926:1)
```

**Comment by dunhuang** [comment]
@woniesong92 did you set Allow Arbitrary Loads = YES in info.plist?

**Comment by sudhirvkumar** [comment]
@dunhuang info.plist is for iOS, how is that going to have any effect in Android? I am confused. @Clcll has mentioned that it works in iOS and it doesn't work in Android...

**Comment by dunhuang** [comment]
@sudhirvkumar  I was thinking @woniesong92 's  issue is running in IOS and now I think I might be wrong. 
As for @Clcll  I think some of his fetch option may not be supported in Android.

**Comment by rcabral90** [comment]
Running into the same issue with Android using axios & fetch.
I am resorting to using a Native Module for Android as a temporary solution.

**Comment by lacker** [comment]
Hey, I think we really don't have enough information here to debug this problem. Stack Overflow might be better for this, that way someone else can help edit the question. I'm going to close this issue, I'd try Stack Overflow if I were you.

**Comment by rcabral90** [comment]
Actually I fixed my problem, I just used the IP address Port of my server that host my API.
There is an issue with the security certificate of my server the android.

**Comment by EllisShen** [comment]
Hay guys, I encountered this issue as well.  What I know so far is that `Fetch` API works on newer Android devices whose is Android 5 or above. The issue now seems happened on an older device including Android 4.1, 4.2 and 4.4 and issues seems happened on RN 0.30 and newer.

**Comment by tuxtux00** [comment]
Hi guys,

I also get this issue randomly with the app installed on my iOS device.
I have allowed Arbitrary Loads in my info.plist file.

As this thread has been closed, is there a topic where we can find help ?

**Comment by divyekhanna** [comment]
@rcabral90 Could you detail your solution a bit? I'm using IP Address and port for the server that hosts my API, but it gives me the `Network Error` about once in 5-7 requests

**Comment by minhcasi** [comment]
I had the same issue on iOS, it is okie for first 4-5 requests but the error occur then, please suggest !

**Comment by StephanPartzsch** [comment]
The thing is, that iOS is running in a simulator and Android is running in an emulator.
The `localhost` is pointing to the environment in which the code is running. The emulator emulates a real device while the simulator is only imitating the device.
Therefore the `localhost` on Android is pointing to the emulated Android device. And not to the machine on which your server is running.
The solution is to replace `localhost` with the IP address of your machine.

**Comment by MHaendel** [comment]
Same problem with a public facing webhost. Network request failed.

**Comment by athmangude** [comment]
Also running on this issue now with devices running on Android version 4.*.* using React Native version 0.40.0

**Comment by ta-dpanga** [comment]
I am facing same problem on react native android emulator. Is there a fix or solution. Please reply.

**Comment by ta-dpanga** [comment]
How to replace localhost with IP address of my machine.

**Comment by yoohan-dex** [comment]
I am also facing that

**Comment by rderwin** [comment]
Use 10.0.2.2 for default AVD and 10.0.3.2 for Genymotion
http://stackoverflow.com/questions/5528850/how-to-connect-localhost-in-android-emulator

**Comment by bexoss** [comment]
+1 how can I fix it?
Even I changed `axios.get('localhost:3000/hello')` to `axios.get('127.0.0.1:3000/hello')`, the problem still left.

**Comment by montrealks** [comment]
I replaced  fetch('http://localhost:3000/api/meetups') with the IP adress of my computer and it worked

**Comment by achakra21** [comment]
use your ip address in place of localhost,It should work

**Comment by sukeshthakare** [comment]
@achakra21 @montrealks  still it not working

**Comment by dan-boa** [comment]
I was also facing the issue and i replaced the localhost with my ip address and it worked.

**Comment by haris-ramzan** [comment]
How to solve this error, This error occur when network connection is lost,  ![screenshot_1499843569](https://user-images.githubusercontent.com/25927759/28406370-b8ec671c-6d49-11e7-812c-60949f0f4dff.png)

**Comment by psahni** [comment]
@StephanPartzsch -  You solved my problem. Thanks a lot. 
For others - Please put the IP address of the target server into your request. 
localhost is conflicting with localhost of android emulator.

All credit goes to StephanPartzsch. Respect for you.

**Comment by Xosmond** [comment]
Same problem with ngrok service. Anybody solve right way?

PD: I can't use IP address, I need a special host on the request.

**Comment by vibhor-pandey** [comment]
I'm facing this problem in Android (iOS I solved it)

`
        let url = "https://some.domain.com";
        let body = null;

        let headers = {
          'Content-Type': 'application/json',
        };

        fetch(url, {
          method: "GET",
          headers: headers,
          body: null
        })
        .then((response) => response.json())
        .then((data) => console.log(data))
        .catch((error) => console.log(error));
`

**Package.json**

`
{
     "react": "16.0.0-alpha.12",
    "react-native": "0.47.1",
}
`


**Error:**

### TypeError: Network request failed
    at XMLHttpRequest.xhr.onerror (fetch.js:441)
    at XMLHttpRequest.dispatchEvent (event-target.js:172)
    at XMLHttpRequest.setReadyState (XMLHttpRequest.js:546)
    at XMLHttpRequest.__didCompleteResponse (XMLHttpRequest.js:381)
    at XMLHttpRequest.js:485
    at RCTDeviceEventEmitter.emit (EventEmitter.js:181)
    at MessageQueue.__callFunction (MessageQueue.js:266)
    at MessageQueue.js:103
    at MessageQueue.__guard (MessageQueue.js:231)
    at MessageQueue.callFunctionReturnFlushedQueue (MessageQueue.js:102)

**Comment by melihberberolu** [comment]
@vibhor-pandey I still get this error. Can u fixed?

**Comment by Tom29** [comment]
@rcabral90 How do i know the ip address of my localhost after connected phone to my computer?

**Comment by vibhor-pandey** [comment]
@rcabral, If you're using Mac, then go to Network Preference > Advance Setting (you can find the IP).

In Windows run command "ipconfig", this will show you the IP

**Comment by vibhor-pandey** [comment]
@melihberberolu  Still I'm finding searching for the solution.

if(I_findSolution === true){
     POST_SOLUTION();
}else if(YOU_findSolution === true){
     POST_SOLUTION();
}

**Comment by melihberberolu** [comment]
@vibhor-pandey I'm changed nginx config ssl_ecdh_curve secp384r1 to ssl_ecdh_curve prime256v1 and fixed

**Comment by vibhor-pandey** [comment]
@melihberberolu  How can I change this SSL?

**Comment by uds214125** [comment]
I have solved the error like Network error  by just putting my local address.  
how to find local address ? 
> ipconfig   
IPv4 will be your local address.
const API_URL ="http://192.168.xx.x:3001/getData";
fetch(`${API_URL}`, {
            method: 'POST',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                phone_no : phone
            })
        })

**Comment by baotoan1905** [comment]
Running into the same issue. +

**Comment by RoboTeddy** [comment]
Per @rderwin's comment, here's how I'm setting things up so they work no matter they environment:

```javascript
import {Platform} from 'react-native'

const SERVER_URL = __DEV__ ?
  Platform.select({
    ios: "http://localhost:3000",
    android: "http://10.0.2.2:3000"
  }) :
  "https://my-production-url.com"
```

**Comment by tomazahlin** [comment]
The only way for us was to use the ip address instead of domain name.

**Comment by blainemuri** [comment]
The problem still persists past localhost as well. iOS seems be working fine for me, but Android is not. Running Android sdk 23 - Marshmallow locally, with a POST request to `https://some-domain/auth`.
When connecting to the remote JS debugging, the problem is solved. It seems like routing it through the browser makes the request work fine. But when remote JS debugging is turned off, I still continue to get `Network Request Failed`.

Running `react-native: 0.45.1`

**Comment by jackson-sandland** [comment]
This is a popular thread. Above answer with 60+ thumbs up did the trick!

**Comment by JackWells** [comment]
Is there an alternative to fetch for react-native which works?  We cannot "substitute the IP" for a real-world domain running on https, because its cert relies on the domain-name to match.  How are people working-around this?  Native-code specific to each device?

**Comment by joelGarcia93** [comment]
yeah it works like a charm!! thanks @StephanPartzsch , we have to change localhost instead your IP address for example 201.165.1.3:port

**Comment by lucascoria** [comment]
Just to cover all the angles but you get the same error if the request is a post and your targeting a get.

**Comment by NxSxD** [comment]
Im facing the issue (Network request failed) on **android**

I get the same error using axios or fetch

This is the fetch request

`        fetch('https://rallycoding.herokuapp.com/api/music_albums')
            .then(response => response.json())
            .then(data => {
                this.setState({ title: 'Loaded!!!' });
            })
            .catch((error) => console.log(error));`

I've been googling around for a couple hours and nothing I have found has helped. 
"react-native": "0.51.0"
"react": "16.0.0"
OS: Ubuntu 16.04
Also tried on MacOS X High Sierra

**Comment by simomau** [comment]
You might want to try putting `'Cache-Control': 'no-cache'` in headers. Fixed similar error for me.

**Comment by waifo** [comment]
How to fix this error can someone clearly suggest it please
I am hitting a url "http://xyz.com" not "http://localhost:xyz/"
I am getting the same error

**Comment by NxSxD** [comment]
My issue was the emulator had a dns server error. I fixed this by appending -dns 8.8.8.8 when running the emulator command.

**Comment by waifo** [comment]
I am not getting dns server error, i am getting Network Error as specified in top of the post

**Comment by waifo** [comment]
can anybody help please..?

**Comment by NxSxD** [comment]
With the dns issue I was having, when making a request, I would get a Network Error as specified above. Make sure your emulator has access to the internet is all I'm saying. Open the browser on the emulator and try to navigate to a webpage. That was simply what was going on with me. Good day.

**Comment by aprilmintacpineda** [comment]
I ran into this issue as well. I can confirm that this is an issue with the emulator and not with react-native.

I think the problem is that, `localhost` refers to the machine which a particular program runs on, that means that `localhost` for the android is the emulator itself so `localhost/api/blabla` is actually making a request to the emulator itself. (Please do @mention me for corrections.)

I fixed it by replacing `http://localhost:9000` with `http://10.0.2.2:9000` as described in [https://developer.android.com/studio/run/emulator-networking.html](https://developer.android.com/studio/run/emulator-networking.html). Basically it **forwards** (I'm not sure if that's the right word for this) the request to the machine where the emulator runs on. However, it only works with `http` and not with `https`, which in my case I was using `https`.

**Comment by sjsn** [comment]
@aprilmintacpineda I had the same issue and got it to work.

You can make an https request on a local server that has an unsigned ssl certificate using [react-native-fetch-blob](https://github.com/wkh237/react-native-fetch-blob) instead of the default `fetch` method and set `{ trusty: true }` as seen [here](https://github.com/wkh237/react-native-fetch-blob/issues/29). 

To make it easy, I abstracted my api logic to an `api.android.js` and `api.ios.js` that dictates whether to use `fetch` or `RNFetchBlob`.

**Comment by manisha-sinha01** [comment]
I have android version 3.0.1 and I am accessing a remote server using fetch function .I checked the internet connection on my simulator but still I'm getting the same error. Please suggest .

**Comment by aprilmintacpineda** [comment]
@sjsn nice! I got your idea, in my case I'm using [axios](https://github.com/axios/axios/). I'll look it up later. Thanks!

**Comment by syamjayaraj** [comment]
@StephanPartzsch I just replaced localhost with my ip address. It is working fine.
But the thing is, the ip address is changing in every restart or changing WiFi connection.

**Comment by pravinraut809** [comment]
Nothing is working for me...

**Comment by monolithed** [comment]
The same problem on Android with image
![img-2018-01-31-00-10-18](https://user-images.githubusercontent.com/803674/35591501-3cdcb382-061b-11e8-94fd-a640b5fb5531.png)

**Comment by aprilmintacpineda** [comment]
@monolithed I think that error is telling you exactly what's wrong. You are giving an `object` as a child of an element, that's not going to work. People will be able to help you more if you show us the code that could be the most relevant and related to that error.

**Comment by monolithed** [comment]
@aprilmintacpineda, nope. 

```ts
let body = new FormData();

body.append('image', {
	uri: image,
	name: 'image',
	type: 'image/jpeg', // This is important for Android!!
});

request = await fetch(url, { body });
```

**Comment by aprilmintacpineda** [comment]
@monolithed  Did you managed to solve it yet? where is that code located in the cycle of your software?

**Comment by monolithed** [comment]
@aprilmintacpineda, the problem has already been solved. That exception has occurred due to I forgot to pass the type fields

**Comment by vanmart** [comment]
how woniesong92 says...  Solution for me was instead of using localhost:port for the API address, actually using an IP address of the server.

**Comment by AmirTahani** [comment]
@vibhor-pandey i had the same problem and i fixed it by chaning server ssl to a paid ssl , 
we were using lets encrypt on server and when i checked ssl on this website https://www.ssllabs.com/ssltest/analyze.html we got grade B for some security reasons! 

by the way i changed ssl and now we have grade A+ and it fixed my problem 
hope this helps someone

**Comment by vibhor-pandey** [comment]
@AmirTahani some updates (including SSL) solved my problem 😄

**Comment by DhruvTekchandani** [comment]
I dont know how far this approach is good, but for me if I go to
Tools -> AVD Manager -> Choose my device and wipe the data.

Then re-run the reac-native run-android, it seems to fetch the results perfectly without any network errors.
But it doesnt make sense why this approach works.

**Comment by alexustinovsm** [comment]
@monolithed tnx mate! You have saved my day!

**Comment by akiramouto** [comment]
in Android  i have a problem same too.
please help i have deadline project in my office.

let url = "https://www.domain.com";  // real domain not localhost
let body = null;

    let headers = {
      'Content-Type': 'application/json',
      'key':'xxxxxxxx',
    };

    fetch(url, {
      method: "GET",
      headers: headers,
      body: null
    })
    .then((response) => response.json())
    .then((data) => console.log(data))
    .catch((error) => console.log(error));
`

Package.json

{ "react": "16.0.0", "react-native": "0.50.0", }

Error:

TypeError: Network request failed
    at XMLHttpRequest.xhr.onerror (C:\HDS\Android\react native\reactNative\mt\node_modules\react-native\node_modules\whatwg-fetch\fetch.js:441)
    at XMLHttpRequest.dispatchEvent (C:\HDS\Android\react native\reactNative\mt\node_modules\event-target-shim\lib\event-target.js:172)
    at XMLHttpRequest.setReadyState (C:\HDS\Android\react native\reactNative\mt\node_modules\react-native\Libraries\Network\XMLHttpRequest.js:554)
    at XMLHttpRequest.__didCompleteResponse (C:\HDS\Android\react native\reactNative\mt\node_modules\react-native\Libraries\Network\XMLHttpRequest.js:387)
    at C:\HDS\Android\react native\reactNative\mt\node_modules\react-native\Libraries\Network\XMLHttpRequest.js:493
    at RCTDeviceEventEmitter.emit (C:\HDS\Android\react native\reactNative\mt\node_modules\react-native\Libraries\vendor\emitter\EventEmitter.js:181)
    at MessageQueue.__callFunction (C:\HDS\Android\react native\reactNative\mt\node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:302)
    at C:\HDS\Android\react native\reactNative\mt\node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:116
    at MessageQueue.__guard (C:\HDS\Android\react native\reactNative\mt\node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:265)
    at MessageQueue.callFunctionReturnFlushedQueue (C:\HDS\Android\react native\reactNative\mt\node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:115)

**Comment by vibhor-pandey** [comment]
@akiramouto there is no any exact statement about this error.

1) Uninstall the App(first clear all the Cached Data of the App from mobile)

2) If 1st Solution doesn't work out, then change the SSL

3) The last solution is to some package version conflicts

**Comment by willwish** [comment]
Please check if you can access google.com in the simulator.
If not, just remove the simulator and install again

**Comment by g13ydson** [comment]
Sounds like a problem with certificate: http://stackoverflow.com/questions/25122287/java-security-cert-certpathvalidatorexception-trust-anchor-for-certification-pa

I tested some sites with this tool [](https://www.digicert.com/help/) And the websites that have the error: _SSL Certificate is not trusted_, are the ones that cause the error.

**Comment by g13ydson** [comment]
Sounds like a problem with certificate. https://stackoverflow.com/questions/48198757/cant-fetch-api-resource

I tested some sites with this tool https://www.digicert.com/help/ And the websites that have the error: SSL Certificate is not trusted, are the ones that cause the error, TypeError: Network request failed

**Comment by goldengecko** [comment]
In my case I had to both specify the dns-server when starting up, which allowed me internet access (could open a webpage in chrome), and also turn on remote JS debugging in my app before it would connect.

The following on Mac:
```
cd ~/Library/Android/sdk/tools
emulator -list-avds
```
Will give you a list of emulators, then you will need to start one up with the DNS specified, e.g.
```
emulator -avd Nexus_7_2012_API_27 -dns-server 10.1.1.1
```
Then in your app Command+M, Enable Remote JS Debugging.

**Comment by DylanVann** [comment]
For iOS you'd still need to change it the actual local ip.

In development mode for iOS React Native gets started on the device with reloading working through the local ip, so the app already has it. Maybe we could get access to it somehow in JS so we can avoid having to do it manually.

**Comment by aprilmintacpineda** [comment]
I'm using iOS simulator from XCode, Android emulator from Android Studio.

on Android it's `10.0.0.2`, so you can do: `http(s)//10.0.0.2:9000/put/other/url/here`.
on iOS it's `localhost`, so you can do `http(s)//localhost:9000/put/other/url/here`.

**Comment by Wowgreat** [comment]
@Clcll 
I found the reason for this error.
when i fetch "https://xxxxx",I get this error.But when i fetch "http://xxxx",this error is fixed.So I guess that by default android do not support HTTPS.
I hope someone could tell me how to fix it.

**Comment by aprilmintacpineda** [comment]
@Wowgreat there are multiple reasons as to why HTTPS would fail, it could be with certificates. I've tested Android and iOS on HTTPS and both worked well, in fact, iOS does not allow HTTP instead it will force you to use HTTPS.

**Comment by Wowgreat** [comment]
@aprilmintacpineda 
Thank you telling me those, I am new one in react-native.It is worth mentioning that few days ago I get the same error on iOS.Finally I fixed by https://github.com/NativeScript/NativeScript/issues/1881 and summarize then write in here https://www.codemanclub.cn/index.php/article/getArticleById/100. 
Hope those is useful for you.

**Comment by abury** [comment]
I'm seeing this as well, however only with a HTTP url.
It seems to work fine with my production URL, however as soon as I switch it to a local server it stops working.

Working: `https://api.*****.net`
Not Working: `http://api.trainer.test` (Hosted locally)

Not really sure what would be happening here. I've inspected traffic with charles and the emulator (genymotion) isn't even attempting to make a connection.

**Comment by saidziani** [comment]
I'm working with React native and Flask REST API, i had the same issue.
And i fixed it by adding: <pre>host= '0.0.0.0'</pre>
<pre>
example:
if __name__ == '__main__':
    app.run(debug=True, host= '0.0.0.0')
</pre>

**Comment by aditi2608** [comment]
Just replace the link's https with http
